KAFKA-7747 Check for truncation after leader changes

mumrah opened the pull request:
After the client detects a leader change we need to check the offset of the current leader for truncation. 



TODO expand on this.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

hachikuji committed:
KAFKA-7747 [WIP]; Check for truncation after leader changes

mumrah committed:
WIP

mumrah committed:
Merge remote-tracking branch 'apache/trunk' into KAFKA-7747

mumrah committed:
After-merge cleanup

mumrah committed:
Checkstyle

mumrah committed:
Include leader epoch in ListOffsetRequest from client

mumrah committed:
Refactor some common code

mumrah committed:
Merge remote-tracking branch 'apache/trunk' into KAFKA-7747

mumrah committed:
Post-merge cleanup

mumrah committed:
Fix a null

mumrah committed:
Merge remote-tracking branch 'apache/trunk' into KAFKA-7747

mumrah committed:
Bit of cleanup and comments

jsancio commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java:
```diff
@@ -754,6 +765,24 @@ public void onSuccess(ClientResponse response, RequestFuture<ListOffsetResult> f
 
+    private RequestFuture<OffsetsForLeaderEpochResponse> sendOffsetsForLeaderEpochRequest(
+            final Node node,
+            final Map<TopicPartition, OffsetsForLeaderEpochRequest.PartitionData> partitionMap) {
```
@mumrah @hachikuji, It looks like this method is not used anywhere. If that is true, how is the client discovering the "largest epoch smaller than the current epoch"?

mumrah replied:
This is called from `Fetcher#prepareFetchRequests`. 

How it works is really the crux of this PR 😄. It checks if the leader has changed (up around L931) by comparing the leader+epoch from latest validated fetch position (stored in the subscription state) with the current leader+epoch according to Metadata. Metadata gets update asynchronously and is what we use as a reference for the cluster state. 

So, if the leader changes, Metadata gets updated and the next time we fetch we will see that the epoch has changed and call `sendOffsetsForLeaderEpochRequest`. This sends an OffsetForLeader request (asynchronously) to the broker and then returns. The partitions transition into a "waiting for validation state" and cannot be fetched from until they are validated.

There is some request batching going on as well to make it a bit more efficient, but slightly confuses things.

In `sendOffsetForLeaderRequest` there is a big callback (`RequestFutureAdaptor#onSuccess`) which does the validation check and either clears the "waiting for validation" state or it truncates (or it throws an error if it can't truncate due to missing reset policy).

jsancio replied:
I checked out this branch and grepped against it:
```
$ rg sendOffsetsForLeaderEpochRequest
clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java
768:    private RequestFuture<OffsetsForLeaderEpochResponse> sendOffsetsForLeaderEpochRequest(
```

mumrah replied:
Ah yea, sorry my mistake. I read that as `sendOffsetForLeaderRequest` which does exist. I'll fix

jsancio commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java:
```diff
@@ -551,14 +556,19 @@ private ListOffsetResult fetchOffsetsByTimes(Map<TopicPartition, Long> timestamp
-                log.trace("Returning fetched records at offset {} for assigned partition {} and update " +
-                        "position to {}", position, partitionRecords.partition, nextOffset);
-                subscriptions.position(partitionRecords.partition, nextOffset);
+                if (partitionRecords.nextFetchOffset > position.offset) {
```
I am sure I am missing something but how can this be true when due to line 560 `if (partitionRecords.nextFetchOffset == position.offset) {`?

jsancio replied:
On a side note, unless the local variable and the field are marked as `final`, it is basically impossible for javac to catch this statically because Java's concurrency model. E.g. any thread can modify both `partitionRecords.nextFetchOffset` and `position.offset`

mumrah replied:
Good catch. Let me look into this... might be something I missed during the merge from trunk.

mumrah replied:
I think the code is correct as-is. The call to `partitionRecords.fetchRecords(maxRecords)` updates the nextFetchOffset to the offset of the last record returned plus one.

mumrah committed:
Remove unused method

mumrah committed:
Renaming some stuff, adding comments, and starting on unit tests

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/Metadata.java:
```diff
@@ -444,4 +445,57 @@ private MetadataRequestAndVersion(MetadataRequest.Builder requestBuilder,
+        Node leader = fetch().leaderFor(tp);
+        if (leader == null)
+            leader = Node.noNode();
+        // TODO there a race here between reading the leader node and reading the epoch? Does it matter?
```
The other methods are synchronized. Any reason not to do that here?

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/Metadata.java:
```diff
@@ -444,4 +445,57 @@ private MetadataRequestAndVersion(MetadataRequest.Builder requestBuilder,
+        public static final LeaderAndEpoch NO_LEADER_OR_EPOCH = new LeaderAndEpoch(Node.noNode(), Optional.empty());
+
+        public final Node leader;
+        public final Optional<Integer> epoch; // TODO shouldn't store an Optional
```
Storing optional is fine IMO. We don't really care about serializability.

mumrah replied:
Heh, this was just a comment to myself really 😁 but yea probably not a big deal

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/Metadata.java:
```diff
@@ -444,4 +445,57 @@ private MetadataRequestAndVersion(MetadataRequest.Builder requestBuilder,
+            if (!update.epoch.isPresent())
+                return true;
+
+            // Otherwise, if the new epoch is larger, then
```
This was probably from my original patch, but this sentence is dangling.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java:
```diff
@@ -502,10 +502,14 @@ public boolean refreshCommittedOffsetsIfNeeded(Timer timer) {
+            final SubscriptionState.FetchPosition position = new SubscriptionState.FetchPosition(
+                    offsetAndMetadata.offset(), offsetAndMetadata.leaderEpoch(), leaderAndEpoch);
+
+            log.debug("Setting offset for partition {} to the committed offset {}", tp, position);
```
This was probably unintentionally reduced to debug level while fixing conflicts.

mumrah replied:
✔️

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java:
```diff
@@ -551,14 +556,19 @@ private ListOffsetResult fetchOffsetsByTimes(Map<TopicPartition, Long> timestamp
-                        "position to {}", position, partitionRecords.partition, nextOffset);
-                subscriptions.position(partitionRecords.partition, nextOffset);
+                if (partitionRecords.nextFetchOffset > position.offset) {
+                    ConsumerMetadata.LeaderAndEpoch leaderAndEpoch = metadata.leaderAndEpoch(partitionRecords.partition);
```
Hmm.. It seems like the leader and epoch might change between when the fetch is sent and when it is received. I think we'd want to use the leader and epoch from the current position rather than updating it here?

mumrah replied:
Ah, right. We only want to update that leader+epoch when we successfully validate an offset.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java:
```diff
@@ -592,9 +602,12 @@ private void resetOffsetIfNeeded(TopicPartition partition, Long requestedResetTi
             log.debug("Skipping reset of partition {} since an alternative reset has been requested", partition);
         } else {
-            log.info("Resetting offset for partition {} to offset {}.", partition, offsetData.offset);
+            ConsumerMetadata.LeaderAndEpoch leaderAndEpoch = metadata.leaderAndEpoch(partition);
```
Similar to my previous comment. When we send a ListOffset or Fetch request, we include the current epoch. The leader verifies that that epoch matches its own before returning a response. The response can only be trusted as representing the state for that epoch. If there is a leader change, we need to repeat validation for the new epoch. Does that make sense?

mumrah replied:
Ok, so for the ListOffset response here, we should only update the position if the returned epoch matches that of the Metadata? And if not, just retry? (via `SubscriptionState#resetFailed`?)

hachikuji replied:
This gets a bit confusing. I think we somehow have to remember which epoch was sent in the ListOffset request. What I think we should be checking is whether the epoch in `FetchPosition.currentLeader` matches what was sent in the request. This would ensure that there was not a leader change while the ListOffset request was inflight.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java:
```diff
@@ -892,34 +903,118 @@ public ListOffsetResult() {
             }
         }
+
+        if (!partitionsToValidate.isEmpty()) {
```
I wonder if it makes sense to split this out as basically a separate stage of `KafkaConsumer.updateFetchPositions`. So we check for any partitions needing reset or validation in the same place.

mumrah replied:
👍

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java:
```diff
@@ -892,34 +903,118 @@ public ListOffsetResult() {
+    private RequestFuture<OffsetsForLeaderEpochResponse> sendOffsetForLeaderEpochRequest(
+            Node node, OffsetsForLeaderEpochRequest.Builder builder) {
+
+        return client.send(node, builder).compose(new RequestFutureAdapter<ClientResponse, OffsetsForLeaderEpochResponse>() {
```
Seems we will raise on any errors? We probably want to handle common errors and just retry internally. Probably all the same errors we handle for ListsOffsets should be handled here.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java:
```diff
@@ -892,34 +903,118 @@ public ListOffsetResult() {
+                            } else {
+                                log.warn("Truncation detected for partition {}, but no reset policy is set", respTopicPartition);
+                                truncationWithoutResetPolicy.put(respTopicPartition, new OffsetAndMetadata(
+                                        currentPosition.offset, currentPosition.currentLeader.epoch, null));
```
Hmm.. Shouldn't this be the offset and epoch from the response? That is the point at which the data we consumed locally diverged from the current leader.

mumrah replied:
Yup, that was an oversight. Will fix

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java:
```diff
@@ -892,34 +903,118 @@ public ListOffsetResult() {
+                            // Try to reset the offset
+                            if (subscriptions.hasDefaultOffsetResetPolicy()) {
+                                log.info("Truncation detected for partition {}, resetting offset", respTopicPartition);
+                                subscriptions.requestOffsetReset(respTopicPartition);
```
I think the point of this is to set the fetch offset to the offset and epoch from the response.

mumrah replied:
Ok, so we detect truncation (leader's end offset is lower than the client's next fetch offset). Then we need to reposition ourselves. Do we consider this a reset of any kind? IOW, should the reset policy even be considered here?

Maybe we simply move to the new end-offset and carry on. But in that case, we'll never hit the LogTruncationException error condition.

mumrah replied:
This logic implemented now is to reset to the end offset from the response iff the consumer has a default reset policy defined (doesn't matter which one). If we detect truncation and no offset reset policy is defined, we raise an error.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java:
```diff
@@ -892,34 +903,118 @@ public ListOffsetResult() {
+                    ConsumerMetadata.LeaderAndEpoch leaderAndEpochForEndOffset = new ConsumerMetadata.LeaderAndEpoch(
+                            Node.noNode(), Optional.of(respEndOffset.leaderEpoch()));
+
+                    if (currentLeaderAndEpoch.isObsoletedBy(leaderAndEpochForEndOffset)) {
```
Again my fault, but we can probably come up with a better name for `isObsoletedBy`. Doesn't read very fluently.

Can you explain the purpose of this check? I think the end offset epoch returned in the response will typically be less than the current epoch. When we receive the response, I think we need to check:

1. Does the partition still need validation for the epoch that was requested? This seems a little tricky because if the partition leadership changes quickly, we may need to validate the new epoch before we have finished validating a previous one.
2. Is the end offset returned from the leader less than the fetch offset? Again we may need to check again whether the fetch offset has changed while this request was inflight.

Initially I was thinking that we needed to do some checking of the epoch returned in the response as well, but now I think the offset may be the only thing. The OffsetForLeaderEpoch API will return the end offset of the largest epoch which is less than or equal to the requested epoch. The epoch returned is only useful for brokers . However it is still helpful to include it in the LogTruncation error that we raise below.

mumrah replied:
So should we just replace the `isObsoletedBy` check with something like `subscriptions.awaitingValidation(respTopicPartition)`?

hachikuji reviewed (commented):
Thanks for the patch. Left a few comments.

mumrah committed:
Feedback from PR

mumrah committed:
Merge remote-tracking branch 'apache/trunk' into KAFKA-7747

mumrah committed:
Move OffsetsForLeaderEpoch code into a new class

mumrah committed:
Rename helper class

mumrah committed:
Rename setResetPending to setNextAllowedRetry

mumrah committed:
Just trying something out

mumrah committed:
Remove extra fetcher class

mumrah committed:
Add some comments/docs

mumrah committed:
Merge branch 'exp-async-client' into KAFKA-7747

mumrah committed:
Merge remote-tracking branch 'apache/trunk' into KAFKA-7747

mumrah committed:
Adding some tests

mumrah committed:
Checkstyle

mumrah commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/OffsetsForLeaderEpochClient.java:
```diff
@@ -0,0 +1,126 @@
+        for (TopicPartition topicPartition : requestData.keySet()) {
+            EpochEndOffset epochEndOffset = response.responses().get(topicPartition);
+            if (epochEndOffset == null) {
+                log.warn("Missing partition {} from response, ignoring", topicPartition);
```
If one of the partitions we requested is missing from the response, should we log and ignore like this, retry the partition, or throw an error? @hachikuji thoughts?

hachikuji replied:
Yeah, good question. I'm always torn in cases like this. Logging and retrying seems reasonable to me.

mumrah committed:
Some cleanup

mumrah committed:
Discard OffsetsForLeaderEpoch response if current epoch changes

mumrah commented:
retest this please

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/Metadata.java:
```diff
@@ -444,4 +445,46 @@ private MetadataRequestAndVersion(MetadataRequest.Builder requestBuilder,
+        Node leader = fetch().leaderFor(tp);
+        if (leader == null)
+            leader = Node.noNode();
+        Optional<Integer> epoch = lastSeenLeaderEpoch(tp);
```
I think the epoch can be updated in isolation without knowing the leader. For example, we may observe an offset commit with a higher epoch than we have metadata for. Are we better off setting the leader to `noNode` in that case?

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java:
```diff
@@ -1508,7 +1510,19 @@ public void commitAsync(final Map<TopicPartition, OffsetAndMetadata> offsets, Of
+
+        acquireAndEnsureOpen();
+        try {
+            log.debug("Seeking to offset {} for partition {}", offset, partition);
```
Should be `info` level to match the other seek APIs?

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/LogTruncationException.java:
```diff
@@ -0,0 +1,50 @@
+
+    /**
+     * Get the truncation offsets for the partitions which were truncated.
+     * This is the first offset which is known to diverge from what the consumer read.
```
The consumer will know when the truncation point can be determined precisely. I am wondering if it is worthwhile trying to communicate this through the API somehow?

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/LogTruncationException.java:
```diff
@@ -0,0 +1,50 @@
+ */
+public class LogTruncationException extends OffsetOutOfRangeException {
+
+    private final Map<TopicPartition, OffsetAndMetadata> offsetOutOfRangePartitions;
```
nit: maybe name should match the getter below? 

By the way, I wasn't too thrilled with the name `truncationOffsets` from the original KIP. What we are returning is the first offset which diverges from what the consumer read. Do you think `divergentOffsets` would be any clearer?

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/Metadata.java:
```diff
@@ -444,4 +445,46 @@ private MetadataRequestAndVersion(MetadataRequest.Builder requestBuilder,
+        return new LeaderAndEpoch(leader, epoch);
+    }
+
+    public static class LeaderAndEpoch {
```
nit: perhaps add a `toString`. Not sure it ends up in any log lines yet, but eventually it will.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncClient.java:
```diff
@@ -0,0 +1,69 @@
+
+public abstract class AsyncClient<T1, Req extends AbstractRequest, Resp extends AbstractResponse, T2> {
+
+    protected final Logger log = LoggerFactory.getLogger(this.getClass());
```
We could get this from `LogContext` as well so that we can differentiate consumer instances.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java:
```diff
@@ -410,22 +418,48 @@ public void resetOffsetsIfNeeded() {
+            SubscriptionState.FetchPosition position = this.subscriptions.position(topicPartition);
+
+            // If the leader or epoch have changed, we need to validate the fetch position
+            if (position != null && !position.safeToFetchFrom(leaderAndEpoch)) {
```
I wonder if it would make sense to push this logic into `SubscriptionState.maybeValidatePosition`? The next check at least seems redundant.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java:
```diff
@@ -410,22 +418,48 @@ public void resetOffsetsIfNeeded() {
+            }
+        });
+
+        validateOffsetsAsync();
```
Probably not a big deal, but we could collect the partitions needing validation in the pass above.

mumrah replied:
Yea, true. I'll do that. It will make it match the reset methods more closely.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java:
```diff
@@ -653,6 +694,85 @@ public void onFailure(RuntimeException e) {
+                    // for the partition. If so, it means we have experienced log truncation and need to reposition
+                    // that partition's offset.
+                    offsetsResult.endOffsets().forEach((respTopicPartition, respEndOffset) -> {
+                        Metadata.LeaderAndEpoch currentLeader = metadata.leaderAndEpoch(respTopicPartition);
```
I would think we'd want to use the epoch in `FetchPosition.currentLeader` for this validation, but maybe we're ensured that they are in sync?

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java:
```diff
@@ -653,6 +694,85 @@ public void onFailure(RuntimeException e) {
+
+                        if (subscriptions.awaitingValidation(respTopicPartition)) {
+                            SubscriptionState.FetchPosition currentPosition = subscriptions.position(respTopicPartition);
+                            if (respEndOffset.endOffset() < currentPosition.offset) {
```
Maybe worth adding a comment here. If `currentPosition.lastFetchEpoch` is equal to `respEndOffset.leaderEpoch`, then the truncation is precise.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java:
```diff
@@ -653,6 +694,85 @@ public void onFailure(RuntimeException e) {
+                                }
+                            } else {
+                                // Offset is fine, clear the validation state
+                                subscriptions.validate(respTopicPartition);
```
Maybe the method should be `completeValidation` or something like that?

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java:
```diff
@@ -653,6 +694,85 @@ public void onFailure(RuntimeException e) {
+                    });
+
+                    if (!truncationWithoutResetPolicy.isEmpty()) {
+                        throw new LogTruncationException(truncationWithoutResetPolicy);
```
If the user ignores this exception and calls `poll()`, will we repeat the validation? That may be reasonable, just making sure I understand.

mumrah replied:
At this point, the subscription state is still in AWAITING_VALIDATION, so yes the next `poll()` call should attempt to revalidate asynchronously and these partitions will be excluded for the time being.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java:
```diff
@@ -653,6 +694,85 @@ public void onFailure(RuntimeException e) {
+
+                @Override
+                public void onFailure(RuntimeException e) {
+                    subscriptions.resetFailed(dataMap.keySet(), time.milliseconds() + retryBackoffMs);
```
nit: maybe `resetFailed` needs a new name since it is used to track inflight validation and reset. Maybe just `requestFailed`?

mumrah replied:
Agreed

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java:
```diff
@@ -711,15 +831,17 @@ public void onFailure(RuntimeException e) {
                 log.debug("Leader for partition {} is unknown for fetching offset", tp);
                 metadata.requestUpdate();
                 partitionsToRetry.add(tp);
+                continue; // TODO still do this?
```
What is the question about?

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java:
```diff
@@ -910,34 +1027,47 @@ public ListOffsetResult() {
                 log.trace("Skipping fetch for partition {} because there is an in-flight request to {}", partition, node);
             } else {
+                // TODO safe to get leader-and-epoch here? Should we use the info in partitionInfoAndEpoch?
+                ConsumerMetadata.LeaderAndEpoch leaderAndEpoch = metadata.leaderAndEpoch(partition);
```
Should we use the leader and epoch from the fetch position?

mumrah replied:
If we do this, then callers of `SubscriptionState#seek(TopicPartition, long)` will be missing the leader Node.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/SubscriptionState.java:
```diff
@@ -505,68 +536,110 @@ public void fireOnAssignment(Set<TopicPartition> assignment) {
 
-        private boolean isMissingPosition() {
-            return !hasValidPosition() && !awaitingReset();
+        private boolean hasPosition() {
```
It's a little confusing to say we have a position while it is still being reset. Maybe it could be `isInitializing` or something like that?

hachikuji reviewed (commented):
Thanks for the updates. I still need to review the test cases, but this is looking pretty good.

mumrah committed:
Adding system test for client offset truncation

mumrah committed:
Feedback from PR

mumrah committed:
Test updates

mumrah committed:
Add validPosition method to subscription state

mumrah committed:
Couple more tests

mumrah committed:
Merge remote-tracking branch 'apache/trunk' into KAFKA-7747

mumrah committed:
Some cleanup, PR feedback

mumrah committed:
More PR feedback and some tests

mumrah committed:
Potential fix for ConsumerCoordinator validating the loaded offset

mumrah committed:
Fix position initialization issue

mumrah committed:
Use the LeaderAndEpoch from metadata, not the position

hachikuji committed:
Always use FetchPosition.currentLeader as request target

mumrah committed:
Use backoff with async validation requests, add isEmpty check for Node in prepareFetchRequest

mumrah committed:
Guard against AWAIT_RESET -> AWAIT_VALIDATION and make sure validation check is run in updateFetchPositions

mumrah committed:
Validate the fetch position inside prepareFetchRequests

This ensures that the position has the most up-to-date leader as possible. This won't
submit the async validation request, but just transition the subscription state.

mumrah committed:
Don't update metadata if a partition has an error

mumrah committed:
Handle some edge cases

mumrah committed:
Cleanup

mumrah committed:
Improve the state machine a bit

mumrah committed:
Fix a metadata response handling edge case

mumrah committed:
Fix error handling in AsyncClient

mumrah committed:
Merge remote-tracking branch 'apache/trunk' into KAFKA-7747

mumrah committed:
Update FetcherTest#testOffsetValidationFencing

mumrah committed:
Merge remote-tracking branch 'apache/trunk' into KAFKA-7747

mumrah committed:
Revert log4j changes

mumrah committed:
checkstyle

mumrah commented:
retest this please

mumrah commented:
retest this please

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java:
```diff
@@ -2206,6 +2226,7 @@ private boolean updateFetchPositions(final Timer timer) {
         // by always ensuring that assigned partitions have an initial position.
         if (coordinator != null && !coordinator.refreshCommittedOffsetsIfNeeded(timer)) return false;
 
+
```
nit: unneeded newline

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java:
```diff
@@ -411,22 +417,45 @@ public void resetOffsetsIfNeeded() {
+        });
+
+        // Collect positions needing validation, with backoff
+        Map<TopicPartition, SubscriptionState.FetchPosition> partitionsToValidate = subscriptions
```
Can we save the second pass over the partitions by doing this collection in the loop above? I'm just thinking about MM-like use cases where the number of partitions could be quite large. A possible optimization is to to cache the metadata update version so that we only bother redoing this check if there has actually been a metadata update.

mumrah replied:
The first pass through all the partitions covers the case of metadata changing, but the second pass through is also used to resubmit the async request with backoff. We could remember the last metadata version seen and avoid unnecessary calls to the first loop.

hachikuji replied:
How about we leave this for a follow-up?

mumrah replied:
Works for me 👍

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java:
```diff
@@ -1658,9 +1675,9 @@ public long position(TopicPartition partition, final Duration timeout) {
-                Long offset = this.subscriptions.position(partition);
-                if (offset != null)
-                    return offset;
+                SubscriptionState.FetchPosition position = this.subscriptions.validPosition(partition);
```
This is interesting. There is always a race with the next leader election for a valid position. Do you think we need to be strict about the timing? I guess if you provide an epoch in seek(), this would be a good way to force validation before fetching.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java:
```diff
@@ -2196,6 +2213,9 @@ private void close(long timeoutMs, boolean swallowException) {
      */
     private boolean updateFetchPositions(final Timer timer) {
+        // If any partitions have been truncated due to a leader change, we need to validate the offsets
+        fetcher.validateOffsetsIfNeeded();
```
Already mentioned, but I think we can be smarter about caching some state to avoid unnecessary work here. Validation is only needed if we do an unprotected seek or a metadata update arrives. Probably this can be left for a follow-up. It's only a concern when the number of partitions and the poll frequency is high.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java:
```diff
@@ -507,10 +508,16 @@ public boolean refreshCommittedOffsetsIfNeeded(Timer timer) {
             entry.getValue().leaderEpoch().ifPresent(epoch -> this.metadata.updateLastSeenEpochIfNewer(entry.getKey(), epoch));
-            this.subscriptions.seek(tp, offset);
+            this.subscriptions.seek(tp, position);
+            this.subscriptions.maybeValidatePosition(tp, leaderAndEpoch);
```
Hmm.. Doesn't this mean we'd only validate the committed offset if there is a change to the current leader and epoch? 

Also, could `leaderAndEpoch` be updated by the call to `updateLastSeenEpochIfNewer` above?

mumrah replied:
In the position I'm creating above, the leader epoch is empty which will cause it to enter validation. This relies on the `FetchPosition#safeToFetchFrom` behavior. If this seems too convoluted, we could add a `seekAndValidate` method or something similar.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java:
```diff
@@ -1535,8 +1549,11 @@ public void seek(TopicPartition partition, OffsetAndMetadata offsetAndMetadata)
+                    offsetAndMetadata.offset(),
+                    offsetAndMetadata.leaderEpoch(),
+                    this.metadata.leaderAndEpoch(partition));
+            this.subscriptions.seek(partition, newPosition);
```
Don't we need to do validation on this new position? Also, did we lose the call to update the last seen epoch?

mumrah replied:
> Also, did we lose the call to update the last seen epoch?

I'll add that back

> Don't we need to do validation on this new position?

I suppose we do, yea. Perhaps another use case of the `seekAndValidate` method proposed above?

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java:
```diff
@@ -654,6 +690,90 @@ public void onFailure(RuntimeException e) {
+                        if (subscriptions.awaitingValidation(respTopicPartition)) {
+                            SubscriptionState.FetchPosition currentPosition = subscriptions.position(respTopicPartition);
+                            Metadata.LeaderAndEpoch currentLeader = currentPosition.currentLeader;
+                            if (!currentLeader.equals(cachedLeaderAndEpochs.get(respTopicPartition))) {
```
The checks above ensure that we are still in the validating phase and that the current leader epoch hasn't changed. I guess it is still possible that both of these are true, but the user has seeked to a different position. Perhaps we can add position to the cached data above?

mumrah replied:
I think it's still okay as long as the position's epoch hasn't changed. What's the side effect if you seek to offset 10 (FETCHING), do validation (AWAIT_VALIDATION), seek to offset 30 (FETCHING), do validation again (AWAIT_VALIDATION), and then get back the OffsetsForLeader response from the first async validation? I think as long as the position's epoch is the same, there isn't a problem. When the second response comes back it will get ignored since we won't be in the right state. WDYT?

hachikuji replied:
I agree that the important thing is that the position's epoch hasn't changed. That and the current leader epoch are the only input to the OffsetsForLeaderEpoch API.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/SubscriptionState.java:
```diff
@@ -345,33 +354,52 @@ public int numAssignedPartitions() {
+        assignedState(tp).position(position);
+    }
+
+    public boolean maybeValidatePosition(TopicPartition tp, Metadata.LeaderAndEpoch leaderAndEpoch) {
```
Should this be named `currentLeaderAndEpoch`?

hachikuji reviewed (commented):
Just a few more comments. I think there were still some minor previous comments which haven't been addressed.

mumrah committed:
Feedback from PR

mumrah committed:
Add seekAndValidate method to SubscriptionState

mumrah committed:
Merge remote-tracking branch 'apache/trunk' into KAFKA-7747

mumrah committed:
Don't check the current epoch when validating a seek

hachikuji reviewed (approved):
LGTM. Thanks for the patch!
